src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -865,65 +865,73 @@ private int ReadBuffer(Span<char> userBuffer, out bool readToUserBuffer)
             return task;
         }
 
-        private async Task<string?> ReadLineAsyncInternal()
+        private Task<string?> ReadLineAsyncInternal()
         {
-            if (_charPos == _charLen && (await ReadBufferAsync(CancellationToken.None).ConfigureAwait(false)) == 0)
+            async Task<string?> ReadInternalFast()
             {
-                return null;
+                if (await ReadBufferAsync(CancellationToken.None).ConfigureAwait(false) == 0)
+                    return null;
+
+                return await ReadInternalLoop().ConfigureAwait(false);
             }
 
-            StringBuilder? sb = null;
 
-            do
+            async Task<string?> ReadInternalLoop()
             {
-                char[] tmpCharBuffer = _charBuffer;
-                int tmpCharLen = _charLen;
-                int tmpCharPos = _charPos;
-                int i = tmpCharPos;
-
+                StringBuilder? sb = null;
                 do
                 {
-                    char ch = tmpCharBuffer[i];
+                    char[] tmpCharBuffer = _charBuffer;
```

Comment:
Maybe is the usage of a `Span<char>` useful here.

Reply:
It's the same as the sync version. I would rather extracted the core logic and wrap it in a sync/async variants. Do we want to do that in that PR? :)

Reply:
> o we want to do that in that PR?

I really like the fact that this PR is very simple to review. If possible, please send other changes in separate PR.

Reply:
I will.